sys/net/gnrc/conn/tcp/gnrc_conn_tcp.c

```diff
@@ -0,0 +1,560 @@
+    xtimer_t usr_tout;
+    int err = 0;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Caller is not the owner of this connection */
+    if (conn->owner != thread_getpid()) {
+        return -EPERM;
+    }
+
+    /* If the connection is passive and the handle is not accepted */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE && !(conn->status & GNRC_TCP_STATUS_ACCEPTED)) {
+        return -EBADF;
+    }
+
+    /* Connection is in illegal State */
+    if (conn->state != ESTABLISHED && conn->state != FIN_WAIT_1
+        && conn->state != FIN_WAIT_2 && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
+
+    /* Setup User Space Timeout */
+    msg_tout.type = MSG_TYPE_USER_TIMEOUT;
+    xtimer_set_msg(&usr_tout, GNRC_TCP_USER_TIMEOUT_DURATION, &msg_tout, thread_getpid());
+
+    while (err == 0) {
+        /* Try to read available data */
+        err = _fsm(conn, CALL_RECV, NULL, data, max_len);
+
+        /* If there was no data: Wait for next packet or until the timeout fires */
+        if (err <= 0) {
+            msg_receive(&msg);
+            switch (msg.type) {
+                case MSG_TYPE_USER_TIMEOUT:
+                    DEBUG("gnrc_conn_tcp : conn_tcp_recv() : USER_TIMEOUT\n");
+                    _fsm(conn, TIMEOUT_USER, NULL, NULL, 0);
+                    err = -ETIME;
+                    break;
+
+                case MSG_TYPE_NOTIFY_USER:
+                    DEBUG("gnrc_conn_tcp : conn_tcp_recv() : MSG_TYPE_NOTIFY_USER\n");
+                    break;
+
+                default:
+                    DEBUG("gnrc_conn_tcp : conn_tcp_recv() : other message type\n");
+            }
+        }
+    }
+
+    /* Cleanup */
+    if(err != -ETIME){
+        xtimer_remove(&usr_tout);
+    }
+
+    /* If this was a passive connection and the connection is closed: re-open */
+    if ((conn->state == CLOSED) && (conn->status & GNRC_TCP_STATUS_PASSIVE)) {
+        _fsm(conn, CALL_OPEN, NULL, NULL, 0);
+    }
+    return err;
+}
+
+int conn_tcp_send(conn_tcp_t *conn, const void *data, size_t len)
+{
+    msg_t msg;
+    msg_t msg_tout;
+    xtimer_t usr_tout;
+    msg_t msg_probe;
+    xtimer_t tout_probe;
+    size_t tout_probe_duration = 0;
+    bool probing = false;
+    int err = 0;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Guard: Caller is not the owner of this connection */
+    if (conn->owner != thread_getpid()) {
+        return -EPERM;
+    }
+
+    /* If the connection is passive and the handle is not accepted */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE && !(conn->status & GNRC_TCP_STATUS_ACCEPTED)) {
+        return -EBADF;
+    }
+
+    /* Guard: Connection is in an illegal state for this operation */
+    if (conn->state != ESTABLISHED && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
```

Comment:
Won't fix. Depricated, conn implementation will be removed after sock_tcp release.